### DISCUSSION TITLE
How can I store KeyCode in scenes?

### DISCUSSION THREAD

**Discussion by Florian0305** [bevyengine/bevy#1334] [Q&A]
Hi everyone 😄 
I'm a beginner in programming, I've been learning Rust and Bevy by myself, and i'm still learning some basics, sorry for any mistakes.

I'm currently trying to store data thanks to Bevy scenes in my video game and I have some trouble storing KeyCode.

I succeeded loading String values in-game with a scene, changing the value of this String in-game, and then storing it in my scene for the next launch.

I used a component like this in order to do it: 

```
#[derive(Reflect, Default)]
#[reflect(Component)]
pub struct Settings {
    pub move_left: String,
    pub move_right: String,
}
```

My goal is to store KeyCodes and not Strings, in order to save the player's settings (example: move_right: KeyCode::D)
So I changed my structure's attributes's type from String to KeyCode. I thought that everything would have worked, but now my #[derive(Reflect)], which allows serialization, deserialization etc does not work and I can't save or load KeyCodes.

Did I miss something? Is it even possible to use KeyCode in scenes? Do I have to implement the way it should behave?
I really tried for hours to understand what is the problem but I can't figure it out 😞, and I've looked at all examples done with Bevy but I still did not succeed with my basic knowledge. I'm really opened to any suggestions ! 😄

**Comment by TheRawMeatball** [comment]
Hmm, maybe you'd be interested in this crate: https://crates.io/crates/kurinji?

**Reply by Florian0305** [reply]
Hey, thanks for you reply! 😄 
I've looked at your crate and it's quite interesting, I may be using it in the future.
However, I'd really like to understand first how can I store KeyCode with scenes, because I would like to store my own enum/struct for my video game, like this one:

```
pub enum GameLevel {
    None,
    Tutorial,
    Level1,
    Level2,
    }
    ```
    
This is very important, because it would allow me to store the current level of the player (meaning the progression of the player), even if he restarts the game or whatever reason.

**Reply by TheRawMeatball** [reply]
Oh, first off, that crate isn't mine - its just a cool crate I found when looking at awesome-bevy. For your question about storing enums in scenes, that might be blocked on https://github.com/bevyengine/bevy/discussions/1306, but I'm not 100% sure.

**Reply by cart** [reply]
KeyCode is just missing a reflect impl. #1306 isn't a hard requirement. It can be worked around by using reflect_value:

```rust
#[derive(Clone, Copy, Eq, PartialEq, Hash, Serialize, Deserialize, Reflect)]
#[reflect_value(Serialize, Deserialize, PartialEq, Hash)]
pub enum KeyCode {
  ...
}
```

**Reply by Florian0305** [reply]
Oh okay thank you! 😄 
Also while doing my scenes tests, I moved all my systems and my struct in a plugin, but when I compiled, I had an error saying I did not register my type with .register_type<>() whereas I've clearly done it in my plugin before adding my systems.
And when I moved back my .register_type and struct to my main there was no problem. So i don't know if it's a bug or just me missing something.

**Reply by cart** [reply]
Hmm maybe its an order of operations problem? Thats very odd.
